Lib/multiprocessing/util.py

```diff
@@ -167,6 +171,165 @@ def _run_after_forkers():
 def register_after_fork(obj, func):
     _afterfork_registry[(next(_afterfork_counter), id(obj), func)] = obj
 
+
+class _WorkQueue:
+    _sentinel = object()
+
+    def __init__(self):
+        self._queue = queue.SimpleQueue()
+        self._lock = threading.RLock()
+        # Note the Condition uses a different lock to avoid a deadlock
+        # when stopping the thread.
+        self._queue_drained_cond = threading.Condition()
+        self._stopped = False
+        self._thread = None
+
+    def __del__(self):
+        self._atexit()
+
+    def ensure_running(self):
+        with self._lock:
+            if self._stopped:
+                return
+            if self._thread is None:
+                self._start_thread()
+
+    def _start_thread(self):
+        thread = threading.Thread(
+            target=self._work_loop,
+            args=(self._queue, self._queue_drained_cond),
+            name='multiprocessing.util._WorkQueue',
+            daemon=True)
+        thread.start()
+        # Avoid setting self._thread before the thread starts successfully.
+        # This should not happen in real life but can happen in
+        # test_concurrent_futures.
+        self._thread = thread
+
+    def _atexit(self):
+        self.stop()
+
+    def _atfork_before(self):
+        import time
+        self._lock.acquire()
+        self._atfork_was_stopped = self._stopped
+        thread_was_running = self._thread is not None
+        # Make sure we don't have a running thread
+        while self._thread is not None:
+            thread = self._stop_unlocked()
+            self._lock.release()
+            try:
+                thread.join()
+            finally:
+                self._lock.acquire()
+        if thread_was_running:
+            # HACK: os.fork() queries the system for the number of running threads.
+            # However, Thread.join() only ensures that the Python thread state
+            # was destroyed, while the system thread could still be running.
+            # Give it time to exit.
+            time.sleep(0.001)
```

Comment:
This is a bit unfortunate. @gpshead

Reply:
Agreed, this is effectively because our join() has never _truely_ been a join... I filed https://github.com/python/cpython/issues/110829 to track that, it is probably time to fix that.